packages/foundry/contracts/hooks/SereneHook.sol

```diff
@@ -0,0 +1,382 @@
+                    // from `amountCalculated`, so we decrease the amount of tokens the Vault will send to the caller.
+                    //
+                    // The preceding swap operation has already credited the original `amountCalculated`. Since we're
+                    // returning `amountCalculated - hookFee` here, it will only register debt for that reduced amount
+                    // on settlement. This call to `sendTo` pulls `hookFee` tokens of `tokenOut` from the Vault to this
+                    // contract, and registers the additional debt, so that the total debts match the credits and
+                    // settlement succeeds.
+                    feeToken = params.tokenOut;
+                    hookAdjustedAmountCalculatedRaw -= hookFee;
+                } else {
+                    // For EXACT_OUT swaps, the `amountCalculated` is the amount of `tokenIn`. The fee must be taken
+                    // from `amountCalculated`, so we increase the amount of tokens the Vault will ask from the user.
+                    //
+                    // The preceding swap operation has already registered debt for the original `amountCalculated`.
+                    // Since we're returning `amountCalculated + hookFee` here, it will supply credit for that increased
+                    // amount on settlement. This call to `sendTo` pulls `hookFee` tokens of `tokenIn` from the Vault to
+                    // this contract, and registers the additional debt, so that the total debts match the credits and
+                    // settlement succeeds.
+                    feeToken = params.tokenIn;
+                    hookAdjustedAmountCalculatedRaw += hookFee;
+                }
+
+                emit HookFeeCharged(address(this), feeToken, hookFee);
+
+                takenFees[params.pool][address(feeToken)] += hookFee;
+                _vault.sendTo(feeToken, address(this), hookFee);
+            }
+        }
+        return (true, hookAdjustedAmountCalculatedRaw);
+    }
+
+    // Alter the swap fee percentage
+    function onComputeDynamicSwapFeePercentage(
+        PoolSwapParams calldata, // params
+        address, // pool
+        uint256 staticSwapFeePercentage
+    ) public view override returns (bool success, uint256 dynamicSwapFeePercentage) {
+        return (true, staticSwapFeePercentage - ((staticSwapFeePercentage * uint256(hookSwapFeePercentage)) / 1e18));
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            QUEST FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Create a quest from the fees taken from the pool
+     * @param pool The pool from which the fees were taken
+     * @param steps The swap steps to convert the fees to the incentive token
+     */
+    function createQuest(address pool, IBatchRouter.SwapPathStep[][] calldata steps) public nonReentrant {
```

Comment:
It's not safe to have the steps be provided since this is a permissionless function. Someone could create a pool that captures 99% of the fees.